crates/pathfinder/src/merkle_tree.rs

```diff
@@ -0,0 +1,1089 @@
+        self.commit_subtree(&mut *self.root.borrow_mut())?;
+        // unwrap is safe as `commit_subtree` will set the hash.
+        let root = self.root.borrow().hash().unwrap();
+        self.storage.increment_ref_count(root)?;
+        Ok(root)
+    }
+
+    /// Persists any changes in this subtree to storage.
+    ///
+    /// This necessitates recursively calculating the hash of, and
+    /// in turn persisting, any changed child nodes. This is necessary
+    /// as the parent node's hash relies on its childrens hashes.
+    ///
+    /// In effect, the entire subtree gets persisted.
+    fn commit_subtree(&self, node: &mut Node) -> anyhow::Result<()> {
+        use Node::*;
+        match node {
+            // Unresolved nodes are already persisted.
+            Unresolved(_) => {}
+            Leaf(hash) => {
+                self.storage
+                    .upsert(*hash, PersistedNode::Leaf)
+                    .context("Failed to insert leaf node")?;
+            }
+            Binary(binary) if binary.hash.is_some() => {}
+            Edge(edge) if edge.hash.is_some() => {}
+            Binary(binary) => {
+                self.commit_subtree(&mut *binary.left.borrow_mut())?;
+                self.commit_subtree(&mut *binary.right.borrow_mut())?;
+                // This will succeed as `commit_subtree` will set the child hashes.
+                binary.calculate_hash();
+                // unwrap is safe as `commit_subtree` will set the hashes.
+                let left = binary.left.borrow().hash().unwrap();
+                let right = binary.right.borrow().hash().unwrap();
+                let persisted_node = PersistedNode::Binary(PersistedBinaryNode { left, right });
+                // unwrap is safe as we just set the hash.
+                self.storage
+                    .upsert(binary.hash.unwrap(), persisted_node)
+                    .context("Failed to insert binary node")?;
+            }
+            Edge(edge) => {
+                self.commit_subtree(&mut *edge.child.borrow_mut())?;
+                // This will succeed as `commit_subtree` will set the child's hash.
+                edge.calculate_hash();
+
+                // unwrap is safe as `commit_subtree` will set the hash.
+                let child = edge.child.borrow().hash().unwrap();
+                let persisted_node = PersistedNode::Edge(PersistedEdgeNode {
+                    path: edge.path.clone(),
+                    child,
+                });
+                // unwrap is safe as we just set the hash.
+                self.storage
+                    .upsert(edge.hash.unwrap(), persisted_node)
+                    .context("Failed to insert edge node")?;
+            }
+        }
+
+        Ok(())
+    }
+
+    /// Sets the value of a key. To delete a key, set the value to [ZERO_HASH].
+    pub fn set(&mut self, key: StarkHash, value: StarkHash) -> anyhow::Result<()> {
+        if value == ZERO_HASH {
```

Comment:
So zero values and non existent values are essentially not distinguished?

Reply:
By "spec" `get(does_not_exist) == ZERO_HASH`.

Reply:
I'm wondering why I see zero value leaves in the python code :thinking:

Reply:
They may only resolve on committing to storage. Or I may have misinterpreted.

There was also a bug where the zero hash was set for all contract storage on first deployment.